File: Ruby/EnemyStarShip.rb
#encoding: utf-8

# +-------------------------------------------------------------------+ #
# | DEEPSPACE (Ruby)                                                  | #
# | Authors: Shao Jie Hu Chen and Mario Megías Mateo                  | #
# +-------------------------------------------------------------------+ #
# | This class represents the Enemy Starship in a battle.             | #
# +-------------------------------------------------------------------+ #

module Deepspace

require_relative "Loot.rb"
require_relative "Damage.rb"
require_relative "ShotResult.rb"
require_relative "EnemyToUI.rb"

    class EnemyStarShip
        
        def initialize(n, a, s, l, d)
            @name        = n
            @ammoPower   = a
            @shieldPower = s
            @loot        = l
            @damage      = d.copy
        end

        def self.newCopy(e)
            new(e.name, e.ammoPower, e.shieldPower, e.loot, e.damage)
        end

        def getUIversion
            EnemyToUI.new(self)
        end

        def fire 
            ammoPower
        end

        def protection
            shieldPower
        end

        attr_reader :ammoPower, :damage, :loot, :name, :shieldPower
        
        def receiveShot(shot)
            if shieldPower < shot
                return ShotResult::DONOTRESIST
            else
                return ShotResult::RESIST
            end
        end

        def to_s
            s = ""
            s += "name: #{name}"
            s += "; ammoPower: #{ammoPower}"
            s += "; shieldPower: #{shieldPower}"
            s += "; loot: " + loot.to_s
            s += "; damage: " + damage.to_s
            s
        end

    end # end of class EnemyStarShip

end # Deepspace

File: Java/src/deepspace/EnemyStarShip.java
package deepspace;

public class EnemyStarShip implements SpaceFighter {
    private float ammoPower;
    private String name;
    private float shieldPower;
    private Loot loot;
    private Damage damage;

    EnemyStarShip(String n, float a, float s, Loot l, Damage d) {
        this.ammoPower   = a;
        this.name        = n;
        this.shieldPower = s;
        this.loot        = l;         // Not usual.
        this.damage      = d.copy();
    }

    EnemyStarShip(EnemyStarShip e) {
        this(e.name, e.ammoPower, e.shieldPower, e.loot, e.damage);
    }

    EnemyToUI getUIversion() {
        return new EnemyToUI(this);
    }

    @Override
    public float fire() {
        return this.ammoPower;
    }

    public float getAmmoPower() {
        return this.ammoPower;
    }

    public Damage getDamage() {
        return this.damage.copy();
    }

    public Loot getLoot() {
        return this.loot; // Not usual.
    }

    public String getName() {
        return new String(this.name);
    }

    public float getShieldPower() {
        return this.shieldPower;
    }

    @Override
    public float protection() {
        return this.shieldPower;
    }

    @Override
    public ShotResult receiveShot(float shot) {
        if (this.shieldPower < shot) {
            return ShotResult.DONOTRESIST;
        } else {
            return ShotResult.RESIST;
        }
    }

    public String toString() {
        String data = "";

        data += "[ammoPower: " + this.ammoPower + ", name: " + this.name + ", shieldPower: " + this.shieldPower;
        data += ", loot: " + this.loot.toString() + ", damage: " + this.damage.toString() + "]";

        return data;
    }
}
